docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+
+
+#### Size limitations
+
+There are limits on the size of the request bodies sent to the kubernetes API server when creating or updating resources: by default ETCD can only accept request bodies which do not exceed 1.5 MiB (this can be configured with the `--max-request-bytes` flag); the kubernetes API Server has a request body limit of 3 MiB which cannot be set from the outside (with a command line flag); the gRPC configuration used by the API server to talk to ETCD has a limit of 2 MiB per request body which cannot be configured from the outside; and `watch` requests have a 16 MiB limit on the buffer used to stream resources.
+
+This means that if `ShootState` is bigger than 1.5 MiB, the ETCD max request bytes will have to be increased. However, there is still an upper limit of 2 MiB imposed by the gRPC configuration.
+
+If `ShootState` exceeds this size limitation it must make use of configmap/secret references to store the state of extension controllers. This is an implementation detail of Gardener and can be done at a later time if necessary as extensions will not be affected.
+
+
+Spliting the `ShootState` into multiple resources could have a positive benefit on performance as the Gardener API Server and Gardener Controller Manager would handle multiple small resources instead of one big resource.
+
+### Gardener extensions changes
+
+All extension controllers which require state migration must save their state in a new `status.state` field and act on an annotation `core.gardener.cloud/operation=restore` in the respective Custom Resources which should trigger a restoration operation instead of reconciliation. A restoration operation means that the extension has to restore its state in the Shoot's namespace on the __Destination Seed__ from the `status.state` field.
+
+As an example: the `Infrastructure` resource must save the terraform state.
+
+```
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Infrastructure
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+Extensions which do not require state migration should set `status.state=nil` in their Custom Resources and trigger a normal reconciliation operation if the CR contains the `core.gardener.cloud/operation=restore` annotation.
+
```

Comment:
Can you also note that, similar to the [contract of the `operation` annotation](https://github.com/gardener/gardener/blob/master/docs/extensions/reconcile-trigger.md) the extension controller has to remove the annotation itself when it's done?